src/session/content/background.rs

```diff
@@ -0,0 +1,397 @@
+            args_builder.set_vec2(5, &p2);
+            args_builder.set_vec2(6, &p3);
+            args_builder.set_vec2(7, &p4);
+
+            gsk::GLShaderNode::new(gradient_shader, bounds, &args_builder.to_args(), &[])
+        }
+
+        fn colors(dark: bool) -> [graphene::Vec3; 4] {
+            use graphene::Vec3;
+            if dark {
+                [
+                    Vec3::new(0.99607843, 0.76862746, 0.5882353),
+                    Vec3::new(0.8666667, 0.42352942, 0.7254902),
+                    Vec3::new(0.5882353, 0.18431373, 0.7490196),
+                    Vec3::new(0.30980393, 0.35686275, 0.8352941),
+                ]
+            } else {
+                [
+                    Vec3::new(0.85882354, 0.8666667, 0.73333335),
+                    Vec3::new(0.41960785, 0.64705884, 0.5294118),
+                    Vec3::new(0.8352941, 0.84705883, 0.5529412),
+                    Vec3::new(0.53333336, 0.72156864, 0.5176471),
+                ]
+            }
+        }
+
+        fn calculate_positions(progress: f32, phase: usize) -> [graphene::Vec2; 4] {
+            static POSITIONS: [(f32, f32); 8] = [
+                (0.80, 0.10),
+                (0.60, 0.20),
+                (0.35, 0.25),
+                (0.25, 0.60),
+                (0.20, 0.90),
+                (0.40, 0.80),
+                (0.65, 0.75),
+                (0.75, 0.40),
+            ];
+
+            let mut points = [graphene::Vec2::new(0.0, 0.0); 4];
+
+            for i in 0..4 {
+                let start = POSITIONS[(i * 2 + phase) % 8];
+                let end = POSITIONS[(i * 2 + phase + 1) % 8];
+
+                fn interpolate(start: f32, end: f32, value: f32) -> f32 {
+                    start + ((end - start) * value)
+                }
+
+                let x = interpolate(start.0, end.0, progress);
+                let y = interpolate(start.1, end.1, progress);
+
+                points[i] = graphene::Vec2::new(x, y);
+            }
+
+            points
+        }
+    }
+}
+
+glib::wrapper! {
+    pub struct Background(ObjectSubclass<imp::Background>)
+        @extends gtk::Widget, adw::Bin;
+}
+
+impl Background {
+    pub fn new() -> Self {
+        glib::Object::new(&[])
+    }
+
+    pub fn animate(&self) {
+        self.emit_by_name::<()>("animate", &[]);
+    }
+
+    fn ensure_shader(&self) {
+        let imp = self.imp();
+        if imp.shaders.borrow().is_none() {
+            let renderer = self.native().unwrap().renderer();
+
+            let sources = [GRADIENT_SHADER, PATTERN_SHADER];
+
+            let shaders: Vec<_> = sources
+                .iter()
+                .flat_map(|source| {
+                    // let bytes = glib::Bytes::from_static(source);
+                    let shader = gsk::GLShader::from_bytes(&source.into());
+                    if let Err(e) = shader.compile(&renderer) {
+                        if e.message() != "The renderer does not support gl shaders" {
```

Comment:
This feels super weird, isn't there a better way to do this? Maybe check for an id?

Reply:
I can get error code from the pointer, but I don't see any getters for it

![image](https://user-images.githubusercontent.com/63008755/218305215-1e366e91-e49d-4cd6-8cbb-9e0a130b8d09.png)

Reply:
I can add that getter in glib-rs later, but I think that string comparsion is more clear here than comparsion with some number

Reply:
Upd: found how to do it correctly